ehrapy/tools/cohort_tracking/_cohort_tracker.py

```diff
@@ -0,0 +1,393 @@
+    """
+
+    def __init__(self, adata: AnnData, columns: Sequence = None, categorical: Sequence = None) -> None:
+        _check_adata_type(adata)
+
+        self.columns = columns if columns is not None else list(adata.obs.columns)
+
+        if columns is not None:
+            _check_columns_exist(adata.obs, columns)
+        if categorical is not None:
+            _check_columns_exist(adata.obs, categorical)
+            if set(categorical).difference(set(self.columns)):
+                raise ValueError("categorical columns must be in the (selected) columns.")
+
+        self._tracked_steps: int = 0
+        self._tracked_text: list = []
+        self._tracked_operations: list = []
+
+        # if categorical columns specified, use them
+        # else, follow tableone's logic
+        self.categorical = (
+            categorical if categorical is not None else _detect_categorical_columns(adata.obs[self.columns])
+        )
+        self._track_t1: list = []
+
+    def __call__(self, adata: AnnData, label: str = None, operations_done: str = None, **tableone_kwargs: dict) -> None:
+        _check_adata_type(adata)
+        _check_columns_exist(adata.obs, self.columns)
+
+        # track a small text with each tracking step, for the flowchart
+        track_text = label if label is not None else f"Cohort {self.tracked_steps}"
+        track_text += "\n (n=" + str(adata.n_obs) + ")"
+        self._tracked_text.append(track_text)
+
+        # track a small text with the operations done
+        self._tracked_operations.append(operations_done)
+        self._tracked_steps += 1
+
+        # track new stuff
+        t1 = TableOne(adata.obs, columns=self.columns, categorical=self.categorical, **tableone_kwargs)
+        self._track_t1.append(t1)
+
+    def _get_cat_dicts(self, table_one, col):
+        cat_pct = {category: [] for category in table_one.cat_table.loc[col].index}
+        for cat in cat_pct.keys():
+            # if tableone does not have the category of this column anymore, set the percentage to 0
+            # for categorized columns (e.g. gender 1.0/0.0), str(cat) helps to avoid considering the category as a float
+            # if (col, str(cat)) in table_one.cat_table["Overall"].index:
+            pct = float(table_one.cat_table["Overall"].loc[(col, str(cat))].split("(")[1].split(")")[0])
+            # else:
+            #    pct = 0
+            cat_pct[cat] = [pct]
+        return pd.DataFrame(cat_pct).T[0]
+
+    def _get_num_dicts(self, table_one, col):
+        return table_one.cont_table["Overall"].loc[(col, "")]
+
+    def reset(self) -> None:
+        """Resets the `CohortTracker` object.
+
+        A full reset of the `CohortTracker` object.
+        """
+        self._tracked_steps = 0
+        self._track_t1 = []
+        self._tracked_text = []
+        self._tracked_operations = []
+
+    @property
+    def tracked_steps(self):
+        """int: Number of tracked steps."""
```

Comment:
```suggestion
        """Number of tracked steps."""
```

Use a return type hint